plugins/out_s3/s3.c

```diff
@@ -2097,6 +2115,103 @@ static void flush_init(void *out_context)
     }
 }
 
+static flb_sds_t s3_format(struct flb_s3 *ctx,
+                           int total_records,
+                           const char *tag, int tag_len,
+                           const void *data, size_t bytes)
+{
+    flb_sds_t tmp = NULL;
+    flb_sds_t chunk = NULL;
+
+    /* Process chunk */
+    if (ctx->log_key) {
+        tmp = flb_pack_msgpack_extract_log_key(ctx, data, bytes);
+        if (!tmp) {
+            flb_plg_error(ctx->ins, "Could not generate records by using 'log_key' content");
+            return NULL;
+        }
+    }
+    else {
+        /*
+         * if 'log_key' is not defined we create a temporary copy of the content since
+         * other parts of the plugin might try to 'release' the buffer which could
+         * generate some issues.
```

Comment:
I'm still struggling to understand this code... what issues will be generated?

Reply:
I don't remember which exact location happens but that fixed the problem. I wrote couple of unit tests to make sure there are no leaks (just pushed other fixes)

Reply:
@edsiper Sorry I need a more definite explanation here, I don't understand why we need to duplicate this memory. Even if there aren't leaks, we still don't want to unnecessarily increase memory usage?

Reply:
The code in `s3_format` has been refactored to avoid unnecessary copies as much as possible, at least as far as I can tell.